log/field.go

```diff
@@ -0,0 +1,201 @@
+package log
+
+import "math"
+
+type fieldType int
+
+const (
+	stringType fieldType = iota
+	boolType
+	intType
+	int32Type
+	uint32Type
+	int64Type
+	uint64Type
+	float32Type
+	float64Type
+	errorType
+	objectType
+	lazyLoggerType
```

Comment:
one other interesting feature in zap was nested fields, which allow logging hierarchical structured data.